parser/src/format/csv.rs

```diff
@@ -367,6 +373,110 @@ impl Iterator for CsvOutput {
     }
 }
 
+// How many of the recent records to consider when trying to decide if
+// we've entered a new region of bad data of the file.
+const ERROR_BUFFER_WINDOW_SIZE: usize = 1000;
+
+/// A decorating iterator that tracks parsing errors and absorbs a specified
+/// rate of errors. If that rate is exceeded, then all the errors encountered
+/// are returned. Should not be polled again once an error is returned.
+#[derive(Debug)]
+struct ParseErrorBuffer<I> {
+    /// The iterator we're wrapping.
+    inner: I,
+    /// The amount of errors we can absorb before halting parsing.
+    threshold: ErrorThreshold,
+    /// The number of records we've seen.
+    total_records: usize,
+    /// TODO
+    errors_in_buffer: usize,
+    /// The most recent rows.
+    buffer: VecDeque<ParseResult>,
+}
+
+impl<I: Iterator<Item = ParseResult>> ParseErrorBuffer<I> {
+    pub fn new(inner: I, threshold: ErrorThreshold) -> Self {
+        let mut buf = Self {
+            inner,
+            threshold,
+            total_records: 0,
+            errors_in_buffer: 0,
+            buffer: VecDeque::with_capacity(ERROR_BUFFER_WINDOW_SIZE),
+        };
+
+        buf.prefill_buffer();
```

Comment:
This feels a little tricky to be within a `new` function, since we generally expect iterators to be lazy. It seems preferable to make it lazily fill the buffer if needed. But it might also work to just rename `new` to something like `fill_buffer`.

Reply:
Yeah, I went back and forth on this a bit. 

On one hand, it doesn't really operate as expected if you forget to call a `fill_buffer` function after `new` but before `next`. The struct is technically initialized, but not _really_ properly setup. 

On the other, as you pointed out, iterators are generally lazy. This one is a bit weird, in that even if we did `prefill_buffer` at the top of `next`, we'd be lazily triggering the read 1000 items. Kind of a mismatch of expectation there too. At some point, I had `prefill_buffer` there, but I moved it since once the buffer was filled the first time, it always preserves that filled buffer until the inner iterator is empty and it starts draining.

On balance, I think I'm leaning towards your point about laziness. That should be easy to swap.